# 3/16  2. Модуль Collections. Counter и defaultdict

## Counter

In [ ]:
# Импортируем объект Counter из модуля collections
from collections import Counter
# Создаём пустой объект Counter
c = Counter()
# Теперь в переменной c хранится объект с возможностями Counter.

# Рассмотрим базовый синтаксис этого инструмента. Например, будем считать цвета проезжающих машин: если встретили красную машину, посчитаем её. Для этого прибавим к ключу 'red' единицу. Синтаксис очень похож на работу со словарём:

c['red'] += 1
print(c)
# Будет напечатано:
# Counter({'red': 1})

Counter({'red': 1})


In [ ]:
cars = ['red', 'blue', 'black', 'black', 'black', 'red', 'blue', 'red', 'white']
# Посчитать значения, конечно, можно и в цикле, используя синтаксис из предыдущего примера:

c = Counter()
for car in cars:
    c[car] += 1

print(c)
# Counter({'red': 3, 'black': 3, 'blue': 2, 'white': 1})

# Однако гораздо проще при создании Counter сразу передать в круглых скобках итерируемый объект, в котором необходимо посчитать значения:

c = Counter(cars)
print(c)
# Counter({'red': 3, 'black': 3, 'blue': 2, 'white': 1})

print(c['black'])
# 3

print(sum(c.values()))
# 9

print(c.values())
# dict_values([3, 2, 3, 1])


Counter({'red': 3, 'black': 3, 'blue': 2, 'white': 1})
Counter({'red': 3, 'black': 3, 'blue': 2, 'white': 1})
3
9
dict_values([3, 2, 3, 1])


**ОПЕРАЦИИ С COUNTER**

In [ ]:
cars_moscow = ['black', 'black', 'white', 'black', 'black', 'white', 'yellow', 'yellow', 'yellow']
cars_spb = ['red', 'black', 'black', 'white', 'white', 'yellow', 'yellow', 'red', 'white']
#Получим для них счётчики:

counter_moscow = Counter(cars_moscow)
counter_spb = Counter(cars_spb)

print(counter_moscow)
print(counter_spb)

# Counter({'black': 4, 'yellow': 3, 'white': 2})
# Counter({'white': 3, 'red': 2, 'black': 2, 'yellow': 2})

Counter({'black': 4, 'yellow': 3, 'white': 2})
Counter({'white': 3, 'red': 2, 'black': 2, 'yellow': 2})


Чтобы узнать, сколько машин разных цветов встретилось в двух городах, можно сложить два исходных счётчика и получить новый счётчик:

In [ ]:

print(counter_moscow + counter_spb)
# Counter({'black': 6, 'white': 5, 'yellow': 5, 'red': 2})
3


Чтобы узнать разницу между объектами Counter, необходимо воспользоваться функцией subtract, которая меняет тот объект, к которому применяется. В примере выше из значений, посчитанных для Москвы, вычитаются значения, посчитанные для Санкт-Петербурга:


In [ ]:

print(counter_moscow)
print(counter_spb)
# Counter({'black': 4, 'yellow': 3, 'white': 2})
# Counter({'white': 3, 'red': 2, 'black': 2, 'yellow': 2})

counter_moscow.subtract(counter_spb)
print(counter_moscow)
# Counter({'black': 2, 'yellow': 1, 'white': -1, 'red': -2})


Заметьте, что белых машин в counter_spb оказалось больше, чем в counter_moscow, поэтому разность отрицательная. Красных машин в moscow вообще не было, а в spb их оказалось сразу две, поэтому разница равна -2. Значения для black и yellow остались положительными, потому что их было больше.

→ Вычитание с учётом отрицательных чисел удобно использовать для сравнения значений в счётчиках: таким образом можно сразу узнать, каких элементов оказалось больше, а каких — меньше.

К сожалению, функцию subtract не всегда бывает удобно использовать для вычитания, так как модифицируется исходный счётчик. Однако аналога у этой функции нет, поскольку вычитание с помощью оператора "-" приводит к другому результату:



In [ ]:
# Пересоздаём счётчики, потому что объект counter_moscow поменял свои значения
# после функции subtract.
counter_moscow = Counter(cars_moscow)
counter_spb = Counter(cars_spb)

print(counter_moscow - counter_spb)
# Counter({'black': 2, 'yellow': 1})

Counter({'black': 2, 'yellow': 1})


**ДОПОЛНИТЕЛЬНЫЕ ФУНКЦИИ**

In [ ]:
#Чтобы получить список всех элементов, которые содержатся в Counter, используется функция elements(). Она возвращает итератор, поэтому, чтобы напечатать все элементы, распакуем их с помощью *:

print(*counter_moscow.elements())
# black black black black white white yellow yellow yellow
#Обратите внимание, что элементы возвращаются в алфавитном порядке, а не в том порядке, в котором их вносили в счётчик.

 #Чтобы получить список уникальных элементов, достаточно воспользоваться функцией list():

print(list(counter_moscow))
# ['black', 'white', 'yellow']

#С помощью функции dict() можно превратить Counter в обычный словарь:

print(dict(counter_moscow))
# {'black': 4, 'white': 2, 'yellow': 3}

#Функция most_common() позволяет получить список из кортежей элементов в порядке убывания их встречаемости:

print(counter_moscow.most_common())
# [('black', 4), ('yellow', 3), ('white', 2)]
#В неё также можно передать значение, которое задаёт желаемое число первых наиболее частых элементов, например, 2:

print(counter_moscow.most_common(2))
# [('black', 4), ('yellow', 3)]

#Наконец, функция clear() позволяет полностью обнулить счётчик:

counter_moscow.clear()
print(counter_moscow)
# Counter()

black black black black white white yellow yellow yellow
['black', 'white', 'yellow']
{'black': 4, 'white': 2, 'yellow': 3}
[('black', 4), ('yellow', 3), ('white', 2)]
[('black', 4), ('yellow', 3)]
Counter()


## DEFAULTDICT

In [ ]:
from collections import defaultdict
groups = defaultdict(list)
students = [('Ivanov',1),('Smirnov',4),('Petrov',3),('Kuznetsova',1),
            ('Nikitina',2),('Markov',3),('Pavlov',2)]
#Обратите внимание, что в скобках мы передаём именно указатель на класс объекта (например list; также можно было бы применить set, dict) без круглых скобок, которые используются для создания нового экземпляра объекта.

#Теперь тот же код, который вызывал ошибку при работе с обычным словарём, сработает так, как ожидается:

for student, group in students:
    groups[group].append(student) # "append" - метод списков и deque!

print(groups)
# defaultdict(<class 'list'>, {1: ['Ivanov', 'Kuznetsova'], 4: ['Smirnov'], 3: ['Petrov', 'Markov'], 2: ['Nikitina', 'Pavlov']})
#В выводе есть небольшое отличие от обычного словаря: печатаются не только элементы словаря, но и само название объекта defaultdict, а также класс объекта, который задан по умолчанию. В данном случае это <class 'list'>.

#Получить элемент из defaultdict по ключу можно так же, как и из обычного словаря:

print(groups[3])
# ['Petrov', 'Markov']

defaultdict(<class 'list'>, {1: ['Ivanov', 'Kuznetsova'], 4: ['Smirnov'], 3: ['Petrov', 'Markov'], 2: ['Nikitina', 'Pavlov']})
['Petrov', 'Markov']


In [ ]:
#Если запрашиваемого ключа нет в словаре, KeyError не возникнет. Вместо этого будет напечатан пустой элемент, который создаётся в словаре по умолчанию:

print(groups[2021])
# []

#Теперь в словаре groups автоматически появился элемент 2021 с пустым списком внутри, несмотря на то что мы его не создавали:

print(groups)
# defaultdict(<class 'list'>, {1: ['Ivanov', 'Kuznetsova'], 4: ['Smirnov'], 3: ['Petrov', 'Markov'], 2: ['Nikitina', 'Pavlov'], 2021: []})

#Итак, вы обратили внимание, что поведение defaultdict в коде отличается от обычного словаря dict. Узнать, с каким именно словарём мы имеем дело в коде, можно с помощью встроенной функции type:

dict_object = dict()
defaultdict_object = defaultdict()

print(type(dict_object))
# <class 'dict'>

print(type(defaultdict_object))
# <class 'collections.defaultdict'>

#Видно, что типы переменных dict_object и defaultdict_object отличаются.

print(dict_object)
# {}

print(defaultdict_object)
# defaultdict(None, {})

#Заметьте, что в этом примере мы не задавали объект, который по умолчанию хранится в экземпляре defaultdict, поэтому объектом по умолчанию является None.

[]
defaultdict(<class 'list'>, {1: ['Ivanov', 'Kuznetsova'], 4: ['Smirnov'], 3: ['Petrov', 'Markov'], 2: ['Nikitina', 'Pavlov'], 2021: []})
<class 'dict'>
<class 'collections.defaultdict'>
{}
defaultdict(None, {})


## ORDEREDDICT

In [ ]:
# Напоминаем способ создания словаря через список кортежей
# (ключ, значение)
data = [('Ivan', 19),('Mark', 25),('Andrey', 23),('Maria', 20)]
client_ages = dict(data)
print(client_ages)
# По результатам 3 повторов получились вот такие результаты:
# {'Maria': 20, 'Mark': 25, 'Ivan': 19, 'Andrey': 23}
# {'Ivan': 19, 'Andrey': 23, 'Mark': 25, 'Maria': 20}
# {'Andrey': 23, 'Mark': 25, 'Maria': 20, 'Ivan': 19}

#Специальный словарь, который гарантирует сохранение ключей в порядке их добавления, называется OrderedDict:

from collections import OrderedDict
data = [('Ivan', 19),('Mark', 25),('Andrey', 23),('Maria', 20)]
ordered_client_ages = OrderedDict(data)
print(ordered_client_ages)

# По результатам 3 повторов получились вот такие результаты:
# OrderedDict([('Ivan', 19), ('Mark', 25), ('Andrey', 23), ('Maria', 20)])
# OrderedDict([('Ivan', 19), ('Mark', 25), ('Andrey', 23), ('Maria', 20)])
# OrderedDict([('Ivan', 19), ('Mark', 25), ('Andrey', 23), ('Maria', 20)])

#Как видите, порядок добавленных ключей сохраняется. Также при выводе словаря на экран печатается тип данных (OrderedDict), к которому он относится.

#Можно, например, отсортировать с помощью функции sorted список кортежей при создании из него OrderedDict, и объекты будут добавлены в порядке сортировки:

data = [('Ivan', 19),('Mark', 25),('Andrey', 23),('Maria', 20)]
# Сортируем по второму значению из кортежа, то есть по возрасту
ordered_client_ages = OrderedDict(sorted(data, key=lambda x: x[1]))
print(ordered_client_ages)
# OrderedDict([('Ivan', 19), ('Maria', 20), ('Andrey', 23), ('Mark', 25)])

{'Ivan': 19, 'Mark': 25, 'Andrey': 23, 'Maria': 20}
OrderedDict([('Ivan', 19), ('Mark', 25), ('Andrey', 23), ('Maria', 20)])
OrderedDict([('Ivan', 19), ('Maria', 20), ('Andrey', 23), ('Mark', 25)])


In [ ]:
#Если теперь добавить нового человека в словарь, новая запись окажется в конце:

ordered_client_ages['Nikita'] = 18
print(ordered_client_ages)
# OrderedDict([('Ivan', 19), ('Maria', 20), ('Andrey', 23), ('Mark', 25), ('Nikita', 18)])

#Если удалить элемент, а затем добавить его снова, он также окажется в конце:

del ordered_client_ages['Andrey']
print(ordered_client_ages)
# OrderedDict([('Ivan', 19), ('Mark', 25), ('Maria', 20), ('Nikita', 18)])

ordered_client_ages['Andrey'] = 23
print(ordered_client_ages)
# OrderedDict([('Ivan', 19), ('Mark', 25), ('Maria', 20), ('Nikita', 18), ('Andrey', 23)])


OrderedDict([('Ivan', 19), ('Maria', 20), ('Andrey', 23), ('Mark', 25), ('Nikita', 18)])
OrderedDict([('Ivan', 19), ('Maria', 20), ('Mark', 25), ('Nikita', 18)])
OrderedDict([('Ivan', 19), ('Maria', 20), ('Mark', 25), ('Nikita', 18), ('Andrey', 23)])


## DEQUE

`from collections import deque`

`dq = deque()`

`print(dq)`

` deque([])`

У deque есть четыре ключевые функции:

append (добавить элемент в конец дека);
appendleft (добавить элемент в начало дека);
pop (удалить и вернуть элемент из конца дека);
popleft (удалить и вернуть элемент из начала дека).

In [ ]:
from collections import deque

clients = deque()
clients.append('Ivanov')
clients.append('Petrov')
clients.append('Smirnov')
clients.append('Tikhonova')
print(clients)
# deque(['Ivanov', 'Petrov', 'Smirnov', 'Tikhonova'])
#Объект deque поддерживает индексацию по элементам:

print(clients[2])
# Smirnov

#Освободилось два оператора — заберём двоих человек из начала очереди с помощью popleft:

first_client = clients.popleft()
second_client = clients.popleft()

print("First client:", first_client)
print("Second client:", second_client)
print(clients)
# First client: Ivanov
# Second client: Petrov
# deque(['Smirnov', 'Tikhonova'])

#Как видите, первые элементы исчезли из очереди. Функции pop и popleft возвращают тот элемент, который они удаляют (последний или первый соответственно).

#Вдруг появился VIP-клиент. Для него тоже нет свободного оператора, но добавить его нужно в начало очереди с помощью appendleft:

clients.appendleft('Vip-client')

print(clients)
# deque(['Vip-client', 'Smirnov', 'Tikhonova'])

#VIP-клиент теперь оказался самым первым в очереди.

#Последний клиент в очереди устал ждать и отменил вызов. Удалим его с помощью pop:

tired_client = clients.pop()
print(tired_client, "left the queue")
print(clients)
# Tikhonova left the queue
# deque(['Vip-client', 'Smirnov'])

#С помощью pop всегда удаляется последний элемент из дэка. Чтобы удалить конкретный элемент по индексу, необходимо воспользоваться встроенной конструкцией del:

clients = deque(['Ivanov', 'Petrov', 'Smirnov', 'Tikhonova'])
print(clients)
# deque(['Ivanov', 'Petrov', 'Smirnov', 'Tikhonova'])

del clients[2]
print(clients)
# deque(['Ivanov', 'Petrov', 'Tikhonova'])

deque(['Ivanov', 'Petrov', 'Smirnov', 'Tikhonova'])
Smirnov
First client: Ivanov
Second client: Petrov
deque(['Smirnov', 'Tikhonova'])
deque(['Vip-client', 'Smirnov', 'Tikhonova'])
Tikhonova left the queue
deque(['Vip-client', 'Smirnov'])
deque(['Ivanov', 'Petrov', 'Smirnov', 'Tikhonova'])
deque(['Ivanov', 'Petrov', 'Tikhonova'])


In [ ]:
#Создадим очередь из клиентов магазинчика на заправке и добавим в неё сразу всех туристов, приехавших на экскурсионном автобусе, с помощью extend:

# В скобках передаём список при создании deque,
# чтобы сразу добавить все его элементы в очередь

shop = deque([1, 2, 3, 4, 5])
print(shop)
# deque([1, 2, 3, 4, 5])

shop.extend([11, 12, 13, 14, 15, 16, 17])
print(shop)
# deque([1, 2, 3, 4, 5, 11, 12, 13, 14, 15, 16, 17])

#Если вдруг у турфирмы имеется договорённость с магазином, что клиенты турфирмы обслуживаются вне очереди, добавим их в начало той же очереди с помощью extendleft:

shop = deque([1, 2, 3, 4, 5])
print(shop)
# deque([1, 2, 3, 4, 5])

shop.extendleft([11, 12, 13, 14, 15, 16, 17])
print(shop)
# deque([17, 16, 15, 14, 13, 12, 11, 1, 2, 3, 4, 5])

deque([1, 2, 3, 4, 5])
deque([1, 2, 3, 4, 5, 11, 12, 13, 14, 15, 16, 17])
deque([1, 2, 3, 4, 5])
deque([17, 16, 15, 14, 13, 12, 11, 1, 2, 3, 4, 5])


### ОЧЕРЕДЬ С ОГРАНИЧЕННОЙ МАКСИМАЛЬНОЙ ДЛИНОЙ

In [ ]:
limited = deque(maxlen=3)
print(limited)
# deque([], maxlen=3)

limited_from_list = deque([1,3,4,5,6,7], maxlen=3)
print(limited_from_list)
# deque([5, 6, 7], maxlen=3)

#Обратите внимание, что теперь дополнительно печатается максимальная длина очереди.

#Также заметьте, что в очереди с ограниченной длиной сохраняются только последние элементы, а первые исчезают из памяти:

limited.extend([1,2,3])
print(limited)
# deque([1, 2, 3], maxlen=3)

print(limited.append(8))
# None

print(limited)
# deque([2, 3, 8], maxlen=3)

#При этом, как видно из результата операции limited.append(8), удаляемый элемент не возвращается, а просто исчезает.

deque([], maxlen=3)
deque([5, 6, 7], maxlen=3)
deque([1, 2, 3], maxlen=3)
None
deque([2, 3, 8], maxlen=3)


In [ ]:
#Ниже приведены средние дневные температуры в Москве за июль:

temps = [20.6, 19.4, 19.0, 19.0, 22.1,
        22.5, 22.8, 24.1, 25.6, 27.0,
        27.0, 25.6, 26.8, 27.3, 22.5,
        25.4, 24.4, 23.7, 23.6, 22.6,
        20.4, 17.9, 17.3, 17.3, 18.1,
        20.1, 22.2, 19.8, 21.3, 21.3,
        21.9]
#Посчитаем динамику средней температуры с усреднением за каждые последние 7 дней для каждого рассматриваемого дня. Для этого воспользуемся очередью с параметром maxlen=7:

days = deque(maxlen=7)

for temp in temps:
    # Добавляем температуру в очередь
    days.append(temp)
    # Если длина очереди оказалась равной максимальной длине очереди (7),
    # печатаем среднюю температуру за последние 7 дней
    if len(days) == days.maxlen:
        print(round(sum(days) / len(days), 2), end='; ')
        #print(days)
# Напечатаем пустую строку, чтобы завершить действие параметра
# end. Иначе следующая строка окажется напечатанной на предыдущей
print("")
# Результат:
# 20.77; 21.27; 22.16; 23.3; 24.44; 24.94; 25.56; 26.2; 25.97;
# 25.94; 25.57; 25.1; 24.81; 24.21; 23.23; 22.57; 21.41; 20.4;
# 19.6; 19.1; 19.04; 18.96; 19.44; 20.01; 20.67;

ДОПОЛНИТЕЛЬНЫЕ ФУНКЦИИ

In [ ]:
#reverse позволяет поменять порядок элементов в очереди на обратный:

dq = deque([1,2,3,4,5])
print(dq)
# deque([1, 2, 3, 4, 5])

dq.reverse()
print(dq)
# deque([5, 4, 3, 2, 1])

#rotate переносит  заданных элементов из конца очереди в начало:

dq = deque([1,2,3,4,5])
print(dq)
# deque([1, 2, 3, 4, 5])

dq.rotate(2)
print(dq)
# deque([4, 5, 1, 2, 3])

#Элементы можно переносить и из начала в конец:

dq = deque([1,2,3,4,5])
print(dq)
# deque([1, 2, 3, 4, 5])

# Отрицательное значение аргумента переносит
# n элементов из начала в конец

dq.rotate(-2)
print(dq)
# deque([3, 4, 5, 1, 2])

#Обратите внимание, что порядок внутри перенесённых элементов остался тем же, каким был изначально. Вспомните, в каком порядке добавляются элементы в начало очереди функцией extend, и сопоставьте с действием rotate.

#Функция index позволяет найти первый индекс искомого элемента, а count позволяет подсчитать, сколько раз элемент встретился в очереди (функции аналогичны одноимённым функциям для списков):

dq = [1,2,4,2,3,1,5,4,4,4,4,4,3]
print(dq.index(4))
# 2

print(dq.count(4))
# 6

#Обратите внимание, что при попытке узнать индекс несуществующего элемента возникнет ValueError:

dq = deque([1,2,4,2,3,1,5,4,4,4,4,4,3])
#print(dq.index(25))
# ValueError: 25 is not in deque

#А вот посчитать несуществующий элемент можно (получится просто 0):

dq = deque([1,2,4,2,3,1,5,4,4,4,4,4,3])
print(dq.count(25))
# 0

#Наконец, функция clear позволяет очистить очередь:

dq = deque([1,2,4,2,3,1,5,4,4,4,4,4,3])
print(dq)
# deque([1, 2, 4, 2, 3, 1, 5, 4, 4, 4, 4, 4, 3])

dq.clear()
print(dq)
# deque([])

deque([1, 2, 3, 4, 5])
deque([5, 4, 3, 2, 1])
deque([1, 2, 3, 4, 5])
deque([4, 5, 1, 2, 3])
deque([1, 2, 3, 4, 5])
deque([3, 4, 5, 1, 2])
2


TypeError: ignored

In [ ]:
def brackets():
    i = 1
    if i:
        clients = deque()
        clients.append('Ivanov')
        clients.append('Petrov')
        clients.append('Smirnov')
        clients.append('Tikhonova')
    print(clients)

brackets()

deque(['Ivanov', 'Petrov', 'Smirnov', 'Tikhonova'])


Задание 3.2 (External resource)

In [ ]:
from collections import OrderedDict

temps = [('2000', -4.4), ('2001', -2.5), ('2002', -4.4), ('2003', -9.5)]

def check(temps):
    wether_dict = OrderedDict(sorted(temps, key=lambda x: x[1], reverse=True))
    return wether_dict

print(check(temps))

OrderedDict([('2001', -2.5), ('2000', -4.4), ('2002', -4.4), ('2003', -9.5)])


Задание 4.3

Напишите функцию brackets(line), которая определяет, является ли последовательность из круглых скобок line правильной.

Примечание 1. Какая последовательность скобок правильная?

Правильной скобочной последовательностью назовём такую последовательность скобок, в которой для каждой открывающей скобки есть последующая соответствующая ей закрывающая скобка. Соответственно, остальные скобочные последовательности назовём неправильными. Пустую строку будем считать правильной последовательностью.

Примечание 2.Для решения этой задачи потребуется использовать стек.

Посимвольно переберите строку. Если встретилась открывающаяся скобка, положите её в стек. Если встретилась закрывающаяся скобка, извлеките скобку из стека.

Если стек пустой, то есть извлечь скобку нельзя, последовательность неправильная.
Если строка закончилась и стек стал пустым, последовательность правильная.
Если в стеке остались скобки, последовательность неправильная.
Пример


print(brackets("(()())"))

 True

print(brackets(""))

 True

print(brackets("(()()))"))

 False

In [ ]:
from collections import deque

def brackets(line):
    # Напишите тело функции
    stack = deque()
    for i in line:
        if i == "(":
            stack.append(i)
        elif i == ")":
            if len(stack) == 0:
                return False
            stack.pop()
    if len(stack) == 0:
        return True
    return False

print(brackets("(()())"))
# True
print(brackets(""))
# True
print(brackets("(()()))"))
# False

True
True
False


Задание 4.4

В переменных center, south и north хранятся списки из перечней купленных позиций в трёх торговых точках, расположенных в разных районах города.
Вначале избавьтесь от излишней вложенности: в каждой переменной (center, south, north) должен храниться объединённый список купленных товаров без разбиения по чекам.

Пример:[['Milk', 'Bread'], ['Meat']] -> ['Milk', 'Bread', 'Meat']

После этого определите, в каком магазине было куплено больше всего товаров.

In [ ]:
from collections import Counter

center = [['Milk', 'Bread'], ['Meat']]
south = ['Milk', 'Bread', 'Meat', 'Meat']
north = [['Milk', 'Bread'], ['Meat'],'Milk', 'Bread', 'Meat', 'Salo']

def bigger_chek(*args):

    stack = Counter()
    for unit in args:
        list_str = str(unit)
        for i in list_str:
            list_str = list_str.replace("[", "").replace("]", "").replace("'", "")
        if unit == center:
            stack["center"] = len((list_str.split(", ")))
        elif unit == south:
            stack["south"] = len((list_str.split(", ")))
        elif unit == north:
            stack["north"] = len((list_str.split(", ")))

    return stack


print(bigger_chek(center, south, north))
# [['Milk', 'Bread'], ['Meat']] -> ['Milk', 'Bread', 'Meat']

Counter({'north': 7, 'south': 4, 'center': 3})


In [ ]:
list_a = [['Milk', 'Bread'], ['Meat']]
list_str = str(list_a)
stack = deque()
for i in list_str:
    list_str = list_str.replace("[", "").replace("]", "").replace("'", "")
print(list_str)
stack.extend(list_str.split(", "))
print(stack)
len(stack)

Milk, Bread, Meat
deque(['Milk', 'Bread', 'Meat'])


3

Задание 4.5

Теперь получите объекты-счётчики (Counter) из каждого полученного в предыдущем задании списка покупок и сохраните их в отдельные переменные (они пригодятся для выполнения следующих задач).

Сколько раз покупали самый редкий товар в магазине north? Запишите ответ в числовой форме.

In [ ]:
from collections import Counter

center = [['Milk', 'Bread'], ['Meat']]
south = ['Milk', 'Bread', 'Meat', 'Meat']
north = [['Milk', 'Bread'], ['Meat'],'Milk', 'Bread', 'Meat', 'Salo']

def bigger_chek(*args):

    stack = Counter()
    for unit in args:
        list_str = str(unit)
        for i in list_str:
            list_str = list_str.replace("[", "").replace("]", "").replace("'", "")
        print(list_str)
        if unit == center:
            stack["center"] = (list_str.split(", "))
            print(stack)
            len(stack)
        elif unit == south:
            stack["south"] = (list_str.split(", "))
        elif unit == north:
            stack["north"] = (list_str.split(", "))

    n = Counter()
    for purchase in stack['north']:
        n[purchase] += 1

    return n.most_common()[-1]

print(bigger_chek(center, south, north))

Milk, Bread, Meat
Counter({'center': ['Milk', 'Bread', 'Meat']})
Milk, Bread, Meat, Meat
Milk, Bread, Meat, Milk, Bread, Meat, Salo
('Salo', 1)



Задание 4.6

Выберите товар, который в магазине center покупали чаще, чем в магазине north:


In [ ]:
from collections import Counter

center = [['Milk', 'Bread'], ['Meat']]
south = ['Milk', 'Bread', 'Meat', 'Meat']
north = [['Milk', 'Bread'], ['Meat'],'Milk', 'Bread', 'Meat', 'Salo']

def bigger_chek(*args):

    stack = Counter()
    for unit in args:
        list_str = str(unit)
        for i in list_str:
            list_str = list_str.replace("[", "").replace("]", "").replace("'", "")
        print(list_str)
        if unit == center:
            stack["center"] = (list_str.split(", "))
            print(stack)
            len(stack)
        elif unit == south:
            stack["south"] = (list_str.split(", "))
        elif unit == north:
            stack["north"] = (list_str.split(", "))

    n = Counter()
    for purchase in stack['north']:
        n[purchase] += 1

        c = Counter()
    for purchase in stack['center']:
        c[purchase] += 1

    f = c - n

    return f

print(bigger_chek(center, south, north))

Milk, Bread, Meat
Counter({'center': ['Milk', 'Bread', 'Meat']})
Milk, Bread, Meat, Meat
Milk, Bread, Meat, Milk, Bread, Meat, Salo
Counter()


Задание 4.7

Есть ли такой товар, который в одном из магазинов покупали чаще, чем в двух других вместе взятых? Если да, выберите магазин с настолько популярным товаром:

In [ ]:
from collections import Counter

center = [['Milk', 'Bread'], ['Meat']]
south = ['Milk', 'Bread', 'Meat', 'Meat']
north = [['Milk', 'Bread'], ['Meat'],'Milk', 'Bread', 'Meat', 'Salo']

def bigger_chek(*args):

    stack = Counter()
    for unit in args:
        list_str = str(unit)
        for i in list_str:
            list_str = list_str.replace("[", "").replace("]", "").replace("'", "")
        print(list_str)
        if unit == center:
            stack["center"] = (list_str.split(", "))
            print(stack)
            len(stack)
        elif unit == south:
            stack["south"] = (list_str.split(", "))
        elif unit == north:
            stack["north"] = (list_str.split(", "))

    n = Counter()
    for purchase in stack['north']:
        n[purchase] += 1

        c = Counter()
    for purchase in stack['center']:
        c[purchase] += 1

        s = Counter()
    for purchase in stack['south']:
        s[purchase] += 1


    moust_c = c - (n + s)
    moust_n = n - (c + s)
    moust_s = s - (n + c)

    return moust_c, moust_n, moust_s

print(bigger_chek(center, south, north))

Milk, Bread, Meat
Counter({'center': ['Milk', 'Bread', 'Meat']})
Milk, Bread, Meat, Meat
Milk, Bread, Meat, Milk, Bread, Meat, Salo
(Counter(), Counter({'Salo': 1}), Counter())


Задание 4.8

Определите суммарное число продаж каждого товара во всех магазинах, сложив все объекты-счётчики. Сколько раз был куплен второй по популярности товар? Запишите ответ в числовой форме.

In [ ]:
from collections import Counter

center = [['Milk', 'Bread'], ['Meat']]
south = ['Milk', 'Bread', 'Meat', 'Meat']
north = [['Milk', 'Bread'], ['Meat'],'Milk', 'Bread', 'Meat', 'Salo']

def bigger_chek(*args):

    stack = Counter()
    for unit in args:
        list_str = str(unit)
        for i in list_str:
            list_str = list_str.replace("[", "").replace("]", "").replace("'", "")
        print(list_str)
        if unit == center:
            stack["center"] = (list_str.split(", "))
            print(stack)
            len(stack)
        elif unit == south:
            stack["south"] = (list_str.split(", "))
        elif unit == north:
            stack["north"] = (list_str.split(", "))

    n = Counter()
    for purchase in stack['north']:
        n[purchase] += 1

        c = Counter()
    for purchase in stack['center']:
        c[purchase] += 1

        s = Counter()
    for purchase in stack['south']:
        s[purchase] += 1


    moust_purched = c + n + s

    return moust_purched.most_common(2)

print(bigger_chek(center, south, north))

Milk, Bread, Meat
Counter({'center': ['Milk', 'Bread', 'Meat']})
Milk, Bread, Meat, Meat
Milk, Bread, Meat, Milk, Bread, Meat, Salo
[('Meat', 5), ('Milk', 4)]


Задание 4.9

Дан список кортежей `ratings` с рейтингами кафе. Кортеж состоит из названия и рейтинга кафе.

Необходимо отсортировать кортеж по убыванию рейтинга. Если рейтинги совпадают, то отсортировать кафе дополнительно по названию в алфавитном порядке.

Получите словарь `cafes` с упорядоченными ключами из отсортированного списка, где ключи — названия кафе, а значения — их рейтинг.


ratings = [('Old York', 3.3), ('New Age', 4.6), ('Old Gold', 3.3), ('General Foods', 4.8),
           ('Belissimo', 4.5), ('CakeAndCoffee', 4.2), ('CakeOClock', 4.2), ('CakeTime', 4.1),
           ('WokToWork', 4.9), ('WokAndRice', 4.9), ('Old Wine Cellar', 3.3), ('Nice Cakes', 3.9)]

In [ ]:
from collections import OrderedDict

ratings = [('Old York', 3.3), ('New Age', 4.6), ('Old Gold', 3.3), ('General Foods', 4.8),
           ('Belissimo', 4.5), ('CakeAndCoffee', 4.2), ('CakeOClock', 4.2), ('CakeTime', 4.1),
           ('WokToWork', 4.9), ('WokAndRice', 4.9), ('Old Wine Cellar', 3.3), ('Nice Cakes', 3.9)]


cafes = OrderedDict(sorted(sorted(ratings), key=lambda x: x[1], reverse=-1))

print(cafes)

OrderedDict([('WokAndRice', 4.9), ('WokToWork', 4.9), ('General Foods', 4.8), ('New Age', 4.6), ('Belissimo', 4.5), ('CakeAndCoffee', 4.2), ('CakeOClock', 4.2), ('CakeTime', 4.1), ('Nice Cakes', 3.9), ('Old Gold', 3.3), ('Old Wine Cellar', 3.3), ('Old York', 3.3)])


Задание 4.10

Напишите функцию task_manager(tasks), которая принимает список задач tasks для нескольких серверов. Каждый элемент списка состоит из кортежа (<номер задачи>, <название сервера>, <высокий приоритет задачи>).

Функция должна создавать словарь и заполнять его задачами по следующему принципу: название сервера — ключ, по которому хранится очередь задач для конкретного сервера.

Если поступает задача без высокого приоритета (последний элемент кортежа — False), добавить номер задачи в конец очереди. Если приоритет высокий, добавить номер в начало.

Для словаря используйте defaultdict, для очереди — deque.

Функция возвращает полученный словарь с задачами.

Пример


tasks = [(36871, 'office', False),
(40690, 'office', False),
(35364, 'voltage', False),
(41667, 'voltage', True),
(33850, 'office', False)]

print(task_manager(tasks))

 defaultdict(, {'voltage': deque([41667, 35364]),

 'office': deque([36871, 40690, 33850])})

In [ ]:
from collections import defaultdict
from collections import deque

tasks = [(36871, 'office', False),
(40690, 'office', False),
(35364, 'voltage', False),
(41667, 'voltage', True),
(33850, 'office', False)]

def task_manager(tasks):
    dict_task = defaultdict(deque)
    for task, server_name, priority in tasks:
        if priority == False:
            dict_task[server_name].append(task)
        else:
            dict_task[server_name].appendleft(task)
    return dict_task

print(task_manager(tasks))
# defaultdict(, {'voltage': deque([41667, 35364]),
# 'office': deque([36871, 40690, 33850])})

defaultdict(<class 'collections.deque'>, {'office': deque([36871, 40690, 33850]), 'voltage': deque([41667, 35364])})


# 5. Модуль NumPy. Типы данных

→ Начнём с целочисленных типов данных в NumPy.

Это тип данных с общим корнем `int`. `Int` может быть со следующими окончаниями: `int8`, `int16`, `int32` и `int64`. Окончание типа данных в NumPy показывает, сколько битов памяти должно быть выделено для хранения переменной.

In [ ]:
import numpy as np
a = np.int8(25)
print(a)
# 25


print(type(a))
# <class 'numpy.int8'>

25
<class 'numpy.int8'>



Чтобы узнать границы int, можно воспользоваться функцией `np.iinfo` (int info):

In [ ]:
# Можно применить к самому
# названию типа данных
np.iinfo(np.int64)
# iinfo(min=-128, max=127, dtype=int8)

# Можно применить к существующему
# конкретному объекту
np.iinfo(a)
# iinfo(min=-128, max=127, dtype=int8)

iinfo(min=-128, max=127, dtype=int8)

В NumPy доступны и беззнаковые целочисленные типы данных. Они имеют корень `uint` (unsigned int — беззнаковое целое). `uint` доступны также с выделением памяти в 8, 16, 32 и 64 бита. При этом максимально возможное число оказывается в два раза больше, чем для соответствующего `int`, поскольку отрицательные числа исключены из типа данных `uint`.

In [ ]:
b = np.uint8(124)
print(b)
# 124
print(type(b))
# <class 'numpy.uint8'>
np.iinfo(b)
# iinfo(min=0, max=255, dtype=uint8)

124
<class 'numpy.uint8'>


iinfo(min=0, max=255, dtype=uint8)

Тип данных не сохранится, если просто присвоить переменной с заданным NumPy-типом данных новое значение:

In [ ]:

a = np.int32(1000)
print(a)
# 1000
print(type(a))
# <class 'numpy.int32'>
a = 2056
print(a)
# 2056
print(type(a))
# <class 'int'>


Вместо этого следует снова указать нужный NumPy-тип данных:

In [ ]:
a = np.int32(1000)
print(a)
# 1000
print(type(a))
# <class 'numpy.int32'>
a = np.int32(2056)
print(a)
# 2056
print(type(a))
# <class 'numpy.int32'>

А вот арифметические операции сохраняют NumPy-тип данных:

In [ ]:
a = np.int32(1000)
b = a + 25
print(b)
# 1025
print(type(b))
# <class 'numpy.int64'>

Если операция проводится с двумя NumPy-типами с фиксированным объёмом памяти, в результате сохраняется наиболее «старший» тип:

In [ ]:
a = np.int32(1000)
b = np.int8(25)
c = a + b
print(c)
# 1025
print(type(c))
# <class 'numpy.int32'>

Следует понимать, что произойдёт, если выделенной памяти для хранения переменной окажется недостаточно.

In [ ]:
a = np.int8(260)
print(a)
# 4
# В переменной a теперь оказалось число 4, а не 260. По сути в переменную записался остаток от деления 260 на 256, а не само число. Ошибка при этом не возникла.

4


Если же при арифметических операциях происходит переполнение максимально выделенной памяти для типа, возникает предупреждение.

In [ ]:
a = np.int32(2147483610)
b = np.int32(2147483605)
print(a, b)
# 2147483610 2147483605
print(a + b)
# -81
# RuntimeWarning: overflow encountered in int_scalars
# Переполнено int'овое значение



In [ ]:
# Чтобы избежать этой ошибки, вначале следовало преобразовать переменные к большему типу:

a = np.int32(2147483610)
b = np.int32(2147483605)
print(a, b)
# 2147483610 2147483605
print(np.int64(a) + np.int64(b))
# 4294967215


### ТИПЫ ДАННЫХ С ПЛАВАЮЩЕЙ ТОЧКОЙ В NUMPY

Их названия строятся по тому же принципу: корень + объём памяти в битах. Беззнаковых `float` нет.

Доступны следующие типы данных `float`: `float16`, `float32`, `float64` (применяется по умолчанию, если объём памяти не задан дополнительно), `float128`.

In [ ]:
# Чтобы узнать границы float и его точность, можно воспользоваться функцией np.finfo(<float тип данных>) (от англ. float info):

np.finfo(np.float16)
# finfo(resolution=0.001, min=-6.55040e+04, max=6.55040e+04, dtype=float16)
np.finfo(np.float32)
# finfo(resolution=1e-06, min=-3.4028235e+38, max=3.4028235e+38, dtype=float32)
np.finfo(np.float64)
# finfo(resolution=1e-15, min=-1.7976931348623157e+308, max=1.7976931348623157e+308, dtype=float64)
np.finfo(np.float128)
# finfo(resolution=1e-18, min=-1.189731495357231765e+4932, max=1.189731495357231765e+4932, dtype=float128)

finfo(resolution=1e-18, min=-1.189731495357231765e+4932, max=1.189731495357231765e+4932, dtype=float128)

Для начала посмотрим на значения min и max. Они указаны в стандартном виде числа. Это такой формат записи числа, при котором число записывается в виде `x*10**n `, где n — целое число, а для  верно: `1 <= x < 10`.

Например, 2021 можно записать в виде `2.021*10**3` . При выводе числа в стандартном виде вместо умножения на 10 в степени  пишется буква , знак степени (+ или -) и сама степень. Следовательно, число 2021 может быть представлено как `2.021e+3`.

*Resolution* (от англ. «разрешение») в выводе finfo означает точность, с которой сохраняется десятичная часть числа в стандартном виде. Для `float16` это 0.001, то есть числа 4.12 и 4.13 будут отличимы друг от друга, а вот 4.124 и 4.125 — нет. Третий знак числа `float16` идёт уже с шагом 0.005:


In [ ]:

print(np.float16(4.12))
# 4.12
print(np.float16(4.13))
# 4.13
print(np.float16(4.123))
# 4.12
print(np.float16(4.124))
# 4.125
print(np.float16(4.125))
# 4.125

4.12
4.13
4.12
4.125
4.125


### ДОПОЛНИТЕЛЬНЫЕ ТИПЫ ДАННЫХ В NUMPY

In [ ]:
print(np.sctypeDict)
print(len(np.sctypeDict))
# 158, но может быть 135 или 139

{'?': <class 'numpy.bool_'>, 0: <class 'numpy.bool_'>, 'byte': <class 'numpy.int8'>, 'b': <class 'numpy.int8'>, 1: <class 'numpy.int8'>, 'ubyte': <class 'numpy.uint8'>, 'B': <class 'numpy.uint8'>, 2: <class 'numpy.uint8'>, 'short': <class 'numpy.int16'>, 'h': <class 'numpy.int16'>, 3: <class 'numpy.int16'>, 'ushort': <class 'numpy.uint16'>, 'H': <class 'numpy.uint16'>, 4: <class 'numpy.uint16'>, 'i': <class 'numpy.int32'>, 5: <class 'numpy.int32'>, 'uint': <class 'numpy.uint64'>, 'I': <class 'numpy.uint32'>, 6: <class 'numpy.uint32'>, 'intp': <class 'numpy.int64'>, 'p': <class 'numpy.int64'>, 7: <class 'numpy.int64'>, 'uintp': <class 'numpy.uint64'>, 'P': <class 'numpy.uint64'>, 8: <class 'numpy.uint64'>, 'long': <class 'numpy.int64'>, 'l': <class 'numpy.int64'>, 'L': <class 'numpy.uint64'>, 'longlong': <class 'numpy.longlong'>, 'q': <class 'numpy.longlong'>, 9: <class 'numpy.longlong'>, 'ulonglong': <class 'numpy.ulonglong'>, 'Q': <class 'numpy.ulonglong'>, 10: <class 'numpy.ulonglong

На самом деле реальных типов данных гораздо меньше, просто одни и те же типы данных могут иметь разные ярлыки. Получить список названий уникальных типов данных NumPy можно с помощью следующего выражения.

In [ ]:
print(*sorted(map(str, set(np.sctypeDict.values()))), sep='\n')

Следует обратить внимание на типы данных bool_ и str_. Они аналогичны bool и str из встроенных в Python, однако записывать их необходимо именно с нижним подчёркиванием, иначе произойдёт приведение к стандартному типу данных, а не типу NumPy. В целом, существенной разницы между этими типами данных нет, однако о такой двойственности следует помнить при сравнении типов переменных: тип bool не является эквивалентным numpy.bool_, несмотря на то что оба типа данных хранят значения True или False.

**Примечание**. В версиях NumPy 1.20 и выше появится предупреждение, если попытаться привести типы с помощью np.bool или np.str, а не np.bool_ или np.str_. Однако в более ранних версиях данное предупреждение не появляется.

In [ ]:
# Пример с bool:

a = True
print(type(a))
# <class 'bool'>

a = np.bool(a)
print(type(a))
# <class 'bool'>

a = np.bool_(a)
print(type(a))
# <class 'numpy.bool_'>

# Значения равны
print(np.bool(True) == np.bool_(True))
# True

# А типы — нет:
print(type(np.bool(True)) == type(np.bool_(True)))
# False

# Пример со str:

a = "Hello world!"
print(type(a))
# <class 'str'>

a = np.str(a)
print(type(a))
# <class 'str'>

a = np.str_(a)
print(type(a))
# <class 'numpy.str_'>

**Небольшое замечание про bool**: несмотря на то что для хранения значения истина/ложь было бы достаточно только одного бита, из-за особенностей работы с памятью компьютера булевая переменная всё равно занимает в памяти целый байт.

## 7/16  6. Модуль NumPy. Массивы

Итак, массив — это структура данных, в которой:

1
Элементы хранятся в указанном порядке.

2
Каждый элемент можно получить по индексу за одинаковое время.

3
Все элементы приведены к одному и тому же типу данных.

4
Максимальное число элементов и объём выделенной памяти заданы заранее.

**Размерностью массива** называют то число индексов, которое необходимо для однозначного получения элемента из массива.

Соответственно, чтобы найти элемент в массиве размерности 1 (строка из чисел) достаточно одного индекса.
В двумерном массиве (таблице из чисел) потребуется уже два индекса: номер строки и номер столбца.
Для трёхмерного массива (например, контейнеры на судне расположены по длине, ширине и высоте судна) потребуется уже три индекса.

**Форма (структура) массива** — это информация о количестве размерностей массива и протяжённости массива по каждой из размерностей. Например, можно задать двумерный массив размера 3x5 — у этой таблицы две размерности: 3 строки и 5 столбцов.

### СОЗДАНИЕ МАССИВА ИЗ СПИСКА

In [ ]:
# оздать массив из списка можно с помощью функции np.array(<объект>):

import numpy as np
arr = np.array([1,5,2,9,10])
arr
# array([ 1,  5,  2,  9, 10])

# Функция np.array возвращает объекты типа numpy.ndarray:

print(type(arr))
# <class 'numpy.ndarray'>
# Название ndarray — это сокращение от n-dimensional array, -мерный массив.

<class 'numpy.ndarray'>


In [ ]:
# Перечислить список из списков можно
# было и в одну строку, но на нескольких
# строках получается нагляднее
nd_arr = np.array([
               [12, 45, 78],
               [34, 56, 13],
               [12, 98, 76]
               ])
nd_arr
# array([[12, 45, 78],
#        [34, 56, 13],
#        [12, 98, 76]])

array([[12, 45, 78],
       [34, 56, 13],
       [12, 98, 76]])

**ТИПЫ ДАННЫХ В МАССИВЕ**

In [ ]:
arr = np.array([1,5,2,9,10])
arr.dtype
# dtype('int64')

dtype('int64')

 Задать тип данных сразу при создании массива можно с помощью параметра `dtype`:



In [ ]:
arr = np.array([1,5,2,9,10], dtype=np.int8)
arr
# array([ 1,  5,  2,  9, 10], dtype=int8)
# При этом тип данных теперь выводится на экран при отображении массива средствами Jupyter Notebook.

array([ 1,  5,  2,  9, 10], dtype=int8)

Теперь, если добавить в arr число больше 127 или меньше -128, оно потеряет исходное значение, как и при преобразовании к меньшему типу:


In [ ]:

arr[2] = 2000
arr
# array([  1,   5, -48,   9,  10], dtype=int8)


Если добавить float в массив int, пропадёт десятичная часть:


In [ ]:

arr[2] = 125.5
arr
# array([  1,   5, 125,   9,  10], dtype=int8)

Строку, которую можно преобразовать в число, можно сразу положить в массив. Она будет приведена к нужному типу автоматически:


In [ ]:

arr[2] = '12'
arr
# array([ 1,  5, 12,  9, 10], dtype=int8)

А вот при попытке положить в массив строку, которую нельзя преобразовать в число, возникнет ошибка:


In [ ]:

arr[2] = 'test'
# ValueError: invalid literal for int() with base 10: 'test'

Смена типа данных

In [ ]:
arr = np.float128(arr)
arr
# array([ 1.,  5., 12.,  9., 10.], dtype=float128)

**СВОЙСТВА NUMPY-МАССИВОВ**

Узнать размерность массива можно с помощью `.ndim`:


In [ ]:
arr = np.array([1,5,2,9,10], dtype=np.int8)
nd_arr = np.array([
               [12, 45, 78],
               [34, 56, 13],
               [12, 98, 76]
               ], dtype=np.int16)


arr.ndim
# 1
nd_arr.ndim
# 2

Узнать общее число элементов в массиве можно с помощью `.size`:


In [ ]:

arr.size
# 5
nd_arr.size
# 9

Форма или структура массива хранится в атрибуте `.shape`:


In [ ]:

arr.shape
# (5,)
nd_arr.shape
# (3, 3)

(3, 3)

Форма массива хранится в виде кортежа с числом элементов, равным размерности массива. Соответственно, для одномерного массива напечатан кортеж длины 1. Обратите внимание, что для двумерного массива вначале было напечатано число «строк», а затем число «столбцов». Это так только отчасти. На самом деле массив как бы состоит из внешних и внутренних массивов: вспомните, что мы передавали список, состоящий из трёх списков, длина каждого из которых равнялась трём. Форма массива определяется от длины внешнего массива (3) к внутреннему (3).

Наконец, узнать, сколько «весит» каждый элемент массива в байтах позволяет
 `.itemsize`:


In [ ]:

arr.itemsize
# 1
nd_arr.itemsize
# 2
# Действительно, в arr хранятся числа в виде int8 (8 бит => 1 байт), а в nd_arr — в виде int16 (16 бит => 2 байта).

2

Можно заранее подготовить массив заданной размерности, заполненный нулями, а потом загружать в него реальные данные по мере необходимости.

Массив из нулей создаётся функцией `np.zeros`. Она принимает аргументы `shape` (обязательный) — форма массива (одно число или кортеж) и `dtype` (необязательный) — тип данных, который будет храниться в массиве.

Создадим одномерный массив из пяти элементов:


In [ ]:

zeros_1d = np.zeros(5)
zeros_1d
# array([0., 0., 0., 0., 0.])

array([0., 0., 0., 0., 0.])

Создадим трёхмерный массив с формой 5x4x3 и типом float32:


In [ ]:
zeros_3d = np.zeros((5,4,3), dtype=np.float32)
print(zeros_3d.shape)
zeros_3d
# (5, 4, 3)

(5, 4, 3)


array([[[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]],

       [[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]],

       [[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]],

       [[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]],

       [[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]]], dtype=float32)


Ещё одной удобной функцией для создания одномерных массивов является arange. Она аналогична встроенной функции range, но обладает рядом особенностей. Вот её сигнатура: `arange([start,] stop, [step,], dtype=None`).

Аргументы `start` (по умолчанию 0), `step` (по умолчанию 1) и `dtype` (определяется автоматически) являются необязательными:

`start` (входит в диапазон возвращаемых значений) задаёт начальное число;

`stop` (не входит в диапазон возвращаемых значений, как и при использовании `range`) задаёт правую границу диапазона;

`step` задаёт шаг, с которым в массив добавляются новые значения.

В отличие от range, в функции arange все перечисленные параметры могут иметь тип `float`.

In [ ]:
# Поэкспериментируем. Создадим массив из пяти чисел от 0 до 4:

np.arange(5)
# array([0, 1, 2, 3, 4])

array([2.5, 3. , 3.5, 4. , 4.5], dtype=float16)

In [ ]:
# Создадим массив от 2.5 до 5:

np.arange(2.5, 5)
# array([2.5, 3.5, 4.5])

In [ ]:
# Создадим массив от 2.5 до 5 с шагом 0.5:

np.arange(2.5, 5, 0.5)
# array([2.5, 3. , 3.5, 4. , 4.5])

In [ ]:
# Создадим массив от 2.5 до 5 с шагом 0.5 и с типом float16:

np.arange(2.5, 5, 0.5, dtype=np.float16)
# array([2.5, 3. , 3.5, 4. , 4.5], dtype=float16)

На самом деле операции с плавающей точкой не всегда бывают предсказуемыми из-за особенностей хранения таких чисел в памяти компьютера. Поэтому для работы с дробными параметрами `start`, `stop` и `step` лучше использовать функцию `linspace` (англ. linear space — линейное пространство). Она тоже возвращает одномерный массив из чисел, расположенных на равном удалении друг от друга между началом и концом диапазона, но обладает немного другим поведением и сигнатурой:

`np.linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None)`

`start` и `stop` являются обязательными параметрами, задающими начало и конец возвращаемого диапазона;

`num` — параметр, задающий число элементов, которое должно оказаться в массиве (по умолчанию 50);

`endpoint` — включён или исключён конец диапазона (по умолчанию включён);

`retstep` (по умолчанию False) позволяет указать, возвращать ли использованный шаг между значениями, помимо самого массива;

`dtype` — уже хорошо знакомый нам параметр, задающий тип данных (если не задан, определяется автоматически).

Давайте потренируемся. Создадим массив из десяти чисел между 1 и 2:

In [ ]:
arr = np.linspace(1, 2, 10)
arr
# array([1.        , 1.11111111, 1.22222222, 1.33333333, 1.44444444,
#        1.55555556, 1.66666667, 1.77777778, 1.88888889, 2.        ])

array([1.        , 1.11111111, 1.22222222, 1.33333333, 1.44444444,
       1.55555556, 1.66666667, 1.77777778, 1.88888889, 2.        ])

Создадим массив из десяти чисел между 1 и 2, не включая 2:

In [ ]:
arr = np.linspace(1, 2, 10, endpoint=False)
arr
# array([1. , 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7, 1.8, 1.9])

array([1. , 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7, 1.8, 1.9])

Узнаем, какой шаг был использован для создания массива из десяти чисел между 1 и 2, где 2 включалось и не включалось:

In [ ]:
arr, step = np.linspace(1, 2, 10, endpoint=True, retstep=True)
print(step)
# 0.1111111111111111

arr, step = np.linspace(1, 2, 10, endpoint=False, retstep=True)
print(step)
print(arr)
# 0.1
# [1.  1.1 1.2 1.3 1.4 1.5 1.6 1.7 1.8 1.9]

0.1111111111111111
0.1
[1.  1.1 1.2 1.3 1.4 1.5 1.6 1.7 1.8 1.9]


Задание 6.6


С каким шагом сгенерируется массив из 60 чисел от -6 до 21 включительно? Ответ округлите до двух знаков после точки.


In [ ]:
number, step = np.linspace(-6, 21, 60, endpoint=True, retstep=True)
print(round(step, 2))

0.46


## ДЕЙСТВИЯ С МАССИВАМИ

Создадим массив из восьми чисел:

In [ ]:
import numpy as np

arr = np.arange(8)
arr
# array([0, 1, 2, 3, 4, 5, 6, 7])

array([0, 1, 2, 3, 4, 5, 6, 7])

Поменять форму массива `arr` можно с помощью присвоения атрибуту `shape` кортежа с желаемой формой:

In [ ]:
arr.shape = (2, 4)
arr
# array([[0, 1, 2, 3],
#        [4, 5, 6, 7]])

array([[0, 1, 2, 3],
       [4, 5, 6, 7]])

Чтобы оставить исходный массив без изменений и дополнительно получить новый массив новой формы, нужно использовать функцию `reshape`. Она также принимает в качестве аргумента кортеж из чисел для формы, но возвращает новый массив, а не изменяет исходный:

In [ ]:
arr = np.arange(8)
arr_new = arr.reshape((2, 4))
arr_new
# array([[0, 1, 2, 3],
#       [4, 5, 6, 7]])

array([[0, 1, 2, 3],
       [4, 5, 6, 7]])

У функции reshape есть дополнительный именованный аргумент `order`. Он задаёт принцип, по которому элементы заполняют массив новой формы. Если `order='C'` (по умолчанию), массив заполняется по строкам, как в примере выше. Если `order='F'`, массив заполняется числами по столбцам:

In [ ]:
arr = np.arange(8)
arr_new = arr.reshape((2, 4), order='F')
arr_new
# array([[0, 2, 4, 6],
#       [1, 3, 5, 7]])

array([[0, 2, 4, 6],
       [1, 3, 5, 7]])

Ещё одной часто используемой операцией с формой массива (особенно двумерного) является **транспонирование**. Эта операция меняет строки и столбцы массива местами. В `NumPy` эту операцию совершает функция `transpose`.

In [ ]:
# Будем работать с двумерным массивом:

arr = np.arange(8)
arr.shape = (2, 4)
arr
# array([[0, 1, 2, 3],
#        [4, 5, 6, 7]])

# Транспонируем его:

arr_trans = arr.transpose()
arr_trans
# array([[0, 4],
#        [1, 5],
#        [2, 6],
#        [3, 7]])

array([[0, 4],
       [1, 5],
       [2, 6],
       [3, 7]])

При транспонировании одномерного массива его форма не меняется:

In [ ]:
arr = np.arange(3)
print(arr.shape)
print(arr)
# (3,)
arr_trans = arr.transpose()
print(arr_trans.shape)
print(arr_trans)
# (3,)

(3,)
[0 1 2]
(3,)
[0 1 2]


### ИНДЕКСЫ И СРЕЗЫ В МАССИВАХ

In [ ]:
# Создадим массив из шести чисел:

arr = np.linspace(1, 2, 6)
arr
# array([1. , 1.2, 1.4, 1.6, 1.8, 2. ])

# Обратиться к его элементу по индексу можно так же, как и к списку:

print(arr[2])
# 1.4

# Привычная запись для срезов работает и для одномерных массивов:

print(arr[2:4])
# [1.4 1.6]

# Наконец, напечатать массив в обратном порядке можно с помощью привычной конструкции [::-1]:

print(arr[::-1])
# [2.  1.8 1.6 1.4 1.2 1. ]

1.4
[1.4 1.6]
[2.  1.8 1.6 1.4 1.2 1. ]


3.0

С многомерными массивами работать немного интереснее. Создадим двумерный массив из одномерного:


In [ ]:
nd_array =  np.linspace(0, 6, 12, endpoint=False).reshape(3,4)
nd_array
# array([[0. , 0.5, 1. , 1.5],
#        [2. , 2.5, 3. , 3.5],
#        [4. , 4.5, 5. , 5.5]])

# Можно воспользоваться привычной записью нескольких индексов в нескольких квадратных скобках:

nd_array[1][2]
# 3.0

3.0

Однако проводить индексацию по массиву в `NumPy` можно проще: достаточно в одних и тех же квадратных скобках перечислить индексы через запятую. Вот так:

In [ ]:
nd_array[1, 2]
# 3.0

3.0

Также через запятую можно передавать срезы или даже их комбинации с индексами. Например, получим все элементы из колонки 3 для первых двух строк:

In [ ]:
nd_array[:2, 2]
# array([1., 3.])
# Несмотря на то что в массиве этот срез является столбцом, вместо него мы получили одномерный массив в виде строки.

Можно применять срезы сразу и к строкам, и к столбцам:

In [ ]:
nd_array[1:, 2:4]
# array([[3. , 3.5],
#       [5. , 5.5]])

array([[3. , 3.5],
       [5. , 5.5]])

Чтобы получить все значения из какой-то оси, можно оставить на её месте двоеточие. Например, из всех строк получим срез с третьего по четвёртый столбцы:

In [ ]:
nd_array[:, 2:4]
# array([[1. , 1.5],
#       [3. , 3.5],
#       [5. , 5.5]])

Чтобы получить самую последнюю ось (в данном случае все столбцы), двоеточие писать необязательно. Строки будут получены целиком по умолчанию:

In [ ]:
nd_array[:2]
# array([[0. , 0.5, 1. , 1.5],
#       [2. , 2.5, 3. , 3.5]])

array([[0. , 0.5, 1. , 1.5],
       [2. , 2.5, 3. , 3.5]])

Задание 7.2

Вам дан массив `mystery`

mystery = np.array([[-13586,  15203,  28445, -27117,  -1781, -17182, -18049],
       [ 25936, -30968,  -1297,  -4593,   6451,  15790,   7181],
       [ 13348,  28049,  28655,  -6012,  21762,  25397,   8225],
       [ 13240,   7994,  32592,  20149,  13754,  11795,   -564],
       [-21725,  -8681,  30305,  22260, -17918,  12578,  29943],
       [-16841, -25392, -17278,  11740,   5916,    -47, -32037]],
      dtype=np.int16)
Выполните все пукты задания по порядку.

Пожалуйста, не меняйте названия переменных, в которых должны появиться ответы.

В переменную `elem_5_3` сохраните элемент из 5 строки и 3 столбца

В переменную `last` сохраните элемент из последней строки последнего столбца

В переменную `line_4` сохраните строку 4

В переменную `col_2` сохраните предпоследний столбец

Из строк 2-4 (включительно) получите столбцы 3-5 (включительно). Результат сохраните в переменную `part`

Сохраните в переменную `rev` последний столбец в обратном порядке

Сохраните в переменную `trans` транспонированный массив

In [ ]:
mystery = np.array([[-13586,  15203,  28445, -27117,  -1781, -17182, -18049],
       [ 25936, -30968,  -1297,  -4593,   6451,  15790,   7181],
       [ 13348,  28049,  28655,  -6012,  21762,  25397,   8225],
       [ 13240,   7994,  32592,  20149,  13754,  11795,   -564],
       [-21725,  -8681,  30305,  22260, -17918,  12578,  29943],
       [-16841, -25392, -17278,  11740,   5916,    -47, -32037]],
      dtype=np.int16)

elem_5_3 = mystery[4, 2]
print(elem_5_3)

ddddd = list(map(lambda x: x-1, mystery.shape))
last = mystery[ddddd[0], ddddd[1]]
print(last)

line_4 = mystery[3]
print(line_4)

col_2 = mystery[:, ddddd[1]-1]
print(col_2)

col_1 = mystery[:, ddddd[1]]
print(col_1)

part = mystery[1:4, 2:5]
print(part)

rev = col_1[::-1]
print(rev)

trans = mystery.transpose()
print(trans)

30305
-32037
[13240  7994 32592 20149 13754 11795  -564]
[-17182  15790  25397  11795  12578    -47]
[-18049   7181   8225   -564  29943 -32037]
[[-1297 -4593  6451]
 [28655 -6012 21762]
 [32592 20149 13754]]
[-32037  29943   -564   8225   7181 -18049]
[[-13586  25936  13348  13240 -21725 -16841]
 [ 15203 -30968  28049   7994  -8681 -25392]
 [ 28445  -1297  28655  32592  30305 -17278]
 [-27117  -4593  -6012  20149  22260  11740]
 [ -1781   6451  21762  13754 -17918   5916]
 [-17182  15790  25397  11795  12578    -47]
 [-18049   7181   8225   -564  29943 -32037]]


СОРТИРОВКА ОДНОМЕРНЫХ МАССИВОВ

Существуют встроенная в `NumPy` функция `sort`. Применять функцию можно двумя способами.

**Способ 1**. Функция `np.sort(<массив>)` возвращает новый отсортированный массив:

In [ ]:
arr = np.array([23,12,45,12,23,4,15,3])
arr_new = np.sort(arr)
print(arr)
# [23 12 45 12 23  4 15  3]
print(arr_new)
# [ 3  4 12 12 15 23 23 45]

[23 12 45 12 23  4 15  3]
[ 3  4 12 12 15 23 23 45]


**Способ 2**. Функция `<массив>.sort()` сортирует исходный массив и возвращает `None`:

In [ ]:
arr = np.array([23,12,45,12,23,4,15,3])
print(arr.sort())
# None
print(arr)
# [ 3  4 12 12 15 23 23 45]

None
[ 3  4 12 12 15 23 23 45]


РАБОТА С ПРОПУЩЕННЫМИ ДАННЫМИ

In [ ]:
# Начнём с примера — создадим массив:

data = np.array([4, 9, -4, 3])

Воспользуемся встроенной в NumPy функцией `sqrt`, чтобы посчитать квадратные корни из элементов.

In [ ]:
roots = np.sqrt(data)
roots
# RuntimeWarning: invalid value encountered in sqrt
# array([2.        , 3.        ,        nan, 1.73205081])

<ipython-input-5-b84b0d3d488d>:1: RuntimeWarning: invalid value encountered in sqrt
  roots = np.sqrt(data)


array([2.        , 3.        ,        nan, 1.73205081])

`nan` - расшифровывается как Not a number (не число). Этот объект аналогичен встроенному типу `None`, но имеет несколько отличий:

**Отличие 1**. `None` является отдельным объектом типа `NoneType`. `np.nan` — это отдельный представитель класса `float`:

In [ ]:
print(type(None))
# <class 'NoneType'>
print(type(np.nan))
# <class 'float'>
type(np.nan)

**Отличие 2**. `None` могут быть равны друг другу, а `np.nan` — нет:

In [ ]:
print(None == None)
# True
print(np.nan == np.nan)
# False

print(None is None)
# True
print(np.nan is np.nan)
# True
print(np.nan is None)
# False

Если попробовать посчитать сумму массива, который содержит `np.nan`, в итоге получится `nan`:

In [ ]:
sum(roots)
# nan

nan

Можно заполнить пропущенные значения, например, нулями. Для этого с помощью функции `np.isnan(<массив>)` узнаем, на каких местах в массиве находятся «не числа»:

In [ ]:
np.isnan(roots)
# array([False, False,  True, False])

array([False, False,  True, False])

Можно использовать полученный массив из `True` и `False` для извлечения элементов из массива roots, на месте которых в булевом массиве указано `True`. Таким способом можно узнать сами элементы, которые удовлетворяют условию `np.isnan`:

In [ ]:
roots[np.isnan(roots)]
# array([nan])

array([nan])

Этим элементам можно присвоить новые значения, например 0:

In [ ]:
roots[np.isnan(roots)] = 0
roots
# array([2.        , 3.        , 0.        , 1.73205081])

array([2.        , 3.        , 0.        , 1.73205081])

In [ ]:
sum(roots)
# 6.732050807568877

6.732050807568877

Задание 7.4

Вам дан массив `mystery`:

mystery = np.array([ 12279., -26024.,  28745.,  np.nan,  31244.,  -2365.,  -6974.,
        -9212., np.nan, -17722.,  16132.,  25933.,  np.nan, -16431.,
        29810.], dtype=np.float32)

Получите булевый массив `nans_index` с информацией о np.nan в массиве mystery: `True` - значение пропущено, `False` - значение не пропущено

В переменную `n_nan` сохраните число пропущенных значений

Скопируйте массив mystery в массив `mystery_new`. Заполните пропущенные значения в массиве `mystery_new` нулями

Поменяйте тип данных в массиве `mystery` на `int32` и сохраните в переменную `mystery_int`

Отсортируйте значения в массиве по возрастанию и сохраните результат в переменную `array`

Сохраните в массив `table` двухмерный массив, полученный из массива `array`. В нём должно быть 5 строк и 3 столбца. Причём порядок заполнения должен быть по столбцам!

Сохраните в переменную 'col' средний столбец из 'table'

In [ ]:
mystery = np.array([ 12279., -26024.,  28745.,  np.nan,  31244.,  -2365.,  -6974.,
        -9212., np.nan, -17722.,  16132.,  25933.,  np.nan, -16431.,
        29810.], dtype=np.float32)

nans_index = np.isnan(mystery)
print(nans_index)

n_nan = len(mystery[np.isnan(mystery)])
print(n_nan)

mystery_new = mystery.copy()
mystery_new[np.isnan(mystery_new)] = 0
print(mystery_new)

mystery_int = np.int32(mystery)
print(mystery_int)

array = np.sort(mystery_int)
print(array)

lenght = len(array)
print(lenght)
table = array.reshape((5, 3), order="F")
print(table)

col = table[:, 1]
print(col)

# 9/16  8. Модуль NumPy. Операции с векторами

В программировании вектором называют одномерный проиндексированный набор данных, другими словами — одномерный массив.

###ВЕКТОРЫ В NUMPY И АРИФМЕТИКА

Произведём сложение двух векторов:

In [ ]:
import numpy as np
vec1 = np.array([2, 4, 7, 2.5])
vec2 = np.array([12, 6, 3.6, 13])
vec1 + vec2
# array([14. , 10. , 10.6, 15.5])

**Для совершения арифметических операций с векторами они должны быть одинаковой длины.**

Поэлементно умножим два вектора одинаковой длины:

In [ ]:
vec1 = np.array([2, 4, 7, 2.5])
vec2 = np.array([12, 6, 3.6])

vec1 * vec2
# ValueError: operands could not be broadcast together with shapes (4,) (3,)
# Ошибка значения: операнд не может быть распространён одновременно на структуры с формами (4,) и (3,).

Исключением является случай, когда операция происходит с вектором и одним числом. Например, вектор целиком можно умножить на число или возвести в степень этого числа:

In [ ]:
vec = np.arange(5)
vec * 10
# array([ 0, 10, 20, 30, 40])
vec ** 2
# array([ 0,  1,  4,  9, 16])

Также векторы можно сравнивать друг с другом поэлементно:

In [ ]:
vec1 = np.array([2, 4, 7, 2.5])
vec2 = np.array([12, 6, 3.6, 13])

vec1 > vec2
# array([False, False,  True, False])

Аналогично можно сравнивать вектор с числом:

In [ ]:
vec = np.array([14,15,9,26,53,5,89])
vec <= 26
# array([ True,  True,  True,  True, False,  True, False])

###ПРОДВИНУТЫЕ ОПЕРАЦИИ С ВЕКТОРАМИ

Посчитаем длину следующего вектора:

In [ ]:
vec = np.array([3, 4])

length = np.sqrt(np.sum(vec ** 2))
print(length)
# 5.0

5.0


Но можно было поступить проще. В `NumPy` есть специальный подмодуль `linalg`, который позволяет производить операции из линейной алгебры.

Для вычисления длины вектора нам потребуется функция `norm`:

In [ ]:
length = np.linalg.norm(vec)
print(length)
# 5.0

5.0


**Расстояние** между двумя векторами, то есть расстояние между их концами, [в евклидовом пространстве] вычисляется как квадратный корень из суммы квадратов разностей соответствующих координат.

 По сути, расстояние между векторами — это длина такого вектора, который является разностью этих векторов. В самом деле, при вычитании двух векторов вычитаются их соответствующие координаты.

In [ ]:
vec1 = np.array([0, 3, 5])
vec2 = np.array([12, 4, 7])
distance = np.sqrt(np.sum((vec1 - vec2) ** 2))
distance
# 12.206555615733702

А теперь применим более простой способ — используем уже известную нам функцию `np.linalg.norm`:

In [ ]:
vec1 = np.array([0, 3, 5])
vec2 = np.array([12, 4, 7])
distance = np.linalg.norm(vec1 - vec2)
distance
# 12.206555615733702

Наконец, **скалярным произведением** двух векторов называют сумму произведений их соответствующих координат.

Откуда такое странное название? Слово «скаляр» — синоним слова «число». То есть результатом вычисления скалярного произведения векторов является число — скаляр. Дело в том, что существуют и другие произведения векторов, не все из которых дают на выходе число.

Реализуем это в коде (по-английски скалярное произведение называют `dot` — точечный — или scalar product, отсюда и такое название переменной):

In [ ]:
vec1 = np.arange(1, 6)
vec2 = np.linspace(10, 20, 5)
scalar_product = np.sum(vec1 * vec2)
scalar_product
# 250.0

250.0

Mожно проще и вообще без формул

Для этого используют функцию `np.dot(x, y)`:



In [ ]:
scalar_product = np.dot(vec1, vec2)
scalar_product
# 250.0

250.0

Если ответ = 0, то вектора перпендикулярны.

В целом, скалярное произведение часто используется для определения угла между векторами.

In [ ]:
x = np.array([25, 0])
y = np.array([0, 10])
np.dot(x, y)
# 0

### БАЗОВЫЕ СТАТИСТИЧЕСКИЕ ФУНКЦИИ ДЛЯ ВЕКТОРОВ

Функции `np.min` и `np.max` позволяют находить максимальное и минимальное значение в векторе. Их можно записывать как в виде `np.min(<vector>)`, так и в виде `<vector>.min()`:

In [ ]:
vec = np.array([2,7,18,28,18,1,8,4])
vec.min()
# 1
np.max(vec)
# 28

1

Функция `mean` позволяет посчитать среднее значение. Больше не требуется реализовывать её «руками»!

In [ ]:
vec.mean()
# 10.75

10.75

Задание 8.9

Найдите медиану массива.

In [ ]:
import numpy as np

mass = np.array([2,7,18,28,18,1,8,4])

# 1-й вариант
print(len(mass))
print(np.sort(mass)[3:5])
print(np.sort(mass)[3:5].mean())

print()

# 2-й вариант
med_even = np.median(mass)

print(med_even)

Задание 8.10

Найдите стандартное отклонение значений в массиве. Ответ округлите до сотых.

In [ ]:
# 1-й вариант
mass = np.array([2,7,18,28,18,1,8,4])
mass_pius = mass - mass.mean()
print(mass.mean())
print(mass_pius)
mass_disp = (np.sum(mass_pius**2))/(len(mass))
print(mass_pius**2)
print(np.sum(mass_pius**2))
print(len(mass))
print(mass_disp)
std_otkl = mass_disp**(1/2)
print(std_otkl)

print()

# 2-й вариант
import numpy as np

res_std = np.std(mass)
print(res_std)

#10/16  9. Модуль NumPy. Случайные числа

###ГЕНЕРАЦИЯ `FLOAT`

Для генерации псевдослучайных чисел в `NumPy` существует подмодуль `random`.

Самой «базовой» функцией в нём можно считать функцию `rand`. По умолчанию она генерирует число с плавающей точкой между 0 (включительно) и 1 (не включительно):

In [ ]:
import numpy as np
np.random.rand()
# 0.06600758835806675

0.275558766185214

Чтобы получить случайное число в диапазоне, например, от 0 до 100, достаточно просто умножить генерируемое число на 100:

In [ ]:
np.random.rand() * 100
# 69.76076924077643

17.442814719734912

получим массив из пяти случайных чисел:

In [ ]:
np.random.rand(5)
# array([0.83745099, 0.58426808, 0.89206204, 0.41149807, 0.42445145])

array([0.18486217, 0.02268617, 0.25187767, 0.84461761, 0.1565656 ])

Массив из двух случайных строк и трёх столбцов:

In [ ]:
np.random.rand(2, 3)
# array([[0.94931212, 0.06680018, 0.26707599],
#      [0.67908873, 0.18001743, 0.97732239]])

array([[0.99681851, 0.50558711, 0.49029962],
       [0.90986077, 0.862851  , 0.7179722 ]])

Функция `rand` может принимать неограниченное число целых чисел для задания формы массива:

In [ ]:
np.random.rand(2, 3, 4, 10, 12, 23)

Если передать в rand кортеж, возникнет ошибка:

In [ ]:
shape = (3, 4)
np.random.rand(shape)
# TypeError: 'tuple' object cannot be interpreted as an integer
# Ошибка типов: кортеж не может быть интерпретирован как целое число.

# Конечно, можно было бы распаковать кортеж, чтобы избавиться от ошибки:

shape = (3, 4)
np.random.rand(*shape)
# array([[0.66169176, 0.19455777, 0.06451088, 0.31919608],
#        [0.73536951, 0.67104408, 0.4762727 , 0.88153576],
#        [0.70672971, 0.96677145, 0.09273995, 0.86356465]])

Но в NumPy есть и другая функция, генерирующая массивы случайных чисел от 0 до 1, которая принимает в качестве аргумента именно кортеж без распаковки. Она называется `sample`:

In [ ]:
shape = (2, 3)
np.random.sample(shape)
# array([[0.39756103, 0.01995168, 0.2768951 ],
#       [0.82195372, 0.26435273, 0.00957881]])

array([[0.33674172, 0.23719888, 0.85913281],
       [0.11945621, 0.36813147, 0.9015283 ]])

Возможно, именно функция sample покажется вам удобнее, поскольку информацию о форме массива обычно удобнее хранить в коде в виде кортежа и не задумываться потом о его распаковке. В остальном функция sample не отличается от rand.

функция uniform:

`uniform(low=0.0, high=1.0, size=None)`

Первые два аргумента — нижняя и верхняя границы диапазона в формате `float`,
третий опциональный аргумент — форма массива (если не задан, возвращается одно число).
Форма массива задаётся кортежем или одним числом.

Запуск без аргументов эквивалентен работе функций `rand` или `sample`:

In [ ]:
np.random.uniform()
# 0.951557685543591

0.4883942018179621

In [ ]:
# Зададим границы диапазона от -30 до 50:

np.random.uniform(-30, 50)
# 38.47365525953661

# Получим пять чисел в интервале от 0.5 до 0.75:

np.random.uniform(0.5, 0.75, size=5)
# array([0.58078945, 0.58860342, 0.73790553, 0.63448265, 0.70920297])

# Получим массив из двух строк и трёх столбцов из чисел в интервале от -1000 до 500:

np.random.uniform(-1000, 500, size=(2, 3))
# array([[ 129.22164163,   77.69090611, -132.9656972 ],
#        [  18.65802226, -317.14793906,   85.3613547 ]])

###ГЕНЕРАЦИЯ INT

 Для генерации целых чисел используется функция random.randint:

`randint(low, high=None, size=None, dtype=int)`

Функцию `randint` нельзя запустить совсем без параметров, необходимо указать хотя бы одно число.

Если указан только аргумент `low`, числа будут генерироваться от 0 до `low`-1, то есть верхняя граница не включается.

Если задать `low` и `high`, числа будут генерироваться от `low` (включительно) до `high` (не включительно).

`size` задаёт форму массива уже привычным для вас образом: одним числом — для одномерного или кортежем — для многомерного.

`dtype` позволяет задать конкретный тип данных, который должен быть использован в массиве.

In [ ]:
# Сгенерируем таблицу 2x3 от 0 до 3 включительно:

np.random.randint(4, size=(2,3))
# array([[3, 0, 1],
#       [2, 1, 3]])
# Чтобы задать и нижнюю, и верхнюю границы самостоятельно, передадим два числа, а затем форму:

np.random.randint(6, 12, size=(3,3))
# array([[ 9,  6, 10],
#        [10, 11, 10],
#        [ 7, 10, 11]])

###ГЕНЕРАЦИЯ ВЫБОРОК

Просто перемешать все числа в массиве позволяет функция random.shuffle.

In [ ]:
arr = np.arange(6)
print(arr)
# [0 1 2 3 4 5]
print(np.random.shuffle(arr))
# None
arr
# array([0, 5, 1, 3, 2, 4])

Чтобы получить новый перемешанный массив, а исходный оставить без изменений, можно использовать функцию `random.permutation`. Она принимает на вход один аргумент — или массив целиком, или одно число:

In [ ]:
playlist = ["The Beatles", "Pink Floyd", "ACDC", "Deep Purple"]
shuffled = np.random.permutation(playlist)
print(shuffled)
# ['The Beatles' 'Pink Floyd' 'Deep Purple' 'ACDC']
print(playlist)
# ['The Beatles', 'Pink Floyd', 'ACDC', 'Deep Purple']

Перемешать набор чисел от 0 до n-1 можно с помощью записи `np.random.permutation(n)`, где `n` — верхняя граница, которая бы использовалась для генерации набора чисел функцией arange.

In [ ]:
np.random.permutation(10)
# array([7, 8, 2, 9, 4, 3, 1, 0, 5, 6])

array([1, 0, 5, 8, 3, 7, 2, 4, 6, 9])

Чтобы получить случайный набор объектов из массива, используется функция `random.choice`:

`choice(a, size=None, replace=True)`

`a` — одномерный массив или число для генерации `arange(a)`;

`size` — желаемая форма массива (число для получения одномерного массива, кортеж — для многомерного; если параметр не задан, возвращается один объект);

`replace` — параметр, задающий, могут ли элементы повторяться (по умолчанию могут).

In [ ]:
workers = ['Ivan', 'Nikita', 'Maria', 'John', 'Kate']

choice = np.random.choice(workers, size=2, replace=False)
print(choice)

['Nikita' 'Kate']


Если попытаться получить без повторений массив большего размера, чем имеется объектов в исходном, возникнет ошибка:

In [ ]:
workers = ['Ivan', 'Nikita', 'Maria', 'John', 'Kate']
choice = np.random.choice(workers, size=10, replace=False)
print(choice)
# ValueError: Cannot take a larger sample than population when 'replace=False'
# Ошибка значения: нельзя получить выборку больше, чем популяция (популяция — весь доступный набор объектов, из которого получаем выборку), если replace=False (то есть выборка без повторений).

Например, получим случайную последовательность, которая образуется в результате десяти подбрасываний игральной кости(с повторениями):

In [ ]:
choice = np.random.choice([1,2,3,4,5,6], size=10)
print(choice)
# [3 5 5 6 6 4 2 2 1 3]

[5 2 1 4 5 5 6 6 1 6]


###SEED ГЕНЕРАТОРА ПСЕВДОСЛУЧАЙНЫХ ЧИСЕЛ

`seed` (от англ. — «зерно»).

Самостоятельно задать seed в `NumPy` можно с помощью функции `np.random.seed(<np.uint32>)`. Число в скобках должно быть в пределах от 0 до 2**32 - 1 (=4294967295).

Зададим `seed` и посмотрим, что получится:

In [ ]:
np.random.seed(23)
np.random.randint(10, size=(3,4))
# array([[3, 6, 8, 9],
#        [6, 8, 7, 9],
#        [3, 6, 1, 2]])

array([[3, 6, 8, 9],
       [6, 8, 7, 9],
       [3, 6, 1, 2]])

Если вы запустите этот код на своём компьютере, то, скорее всего, увидите тот же самый набор чисел!

In [ ]:
np.random.seed(100)
print(np.random.randint(10, size=3))
# [8 8 3]
print(np.random.randint(10, size=3))
# [7 7 0]
print(np.random.randint(10, size=3))
# [4 2 5]

[8 8 3]
[7 7 0]
[4 2 5]


Задание 9.6

Задайте `seed` = 2021.

Сохраните в переменные необходимые значения.

Примечание 1. Не меняйте названия переменных и последовательность генерации случайных чисел в задании.

Примечание 2. Не забудьте импортировать numpy и сразу задать `seed` = 2021

В `simple` сохраните случайное число в диапазоне от 0 до 1

Сгенерируйте 120 чисел в диапазоне от -150 до 2021, сохраните их в переменную `randoms`

Получите массив из случайных целых чисел от 1 до 100 (включительно) из 3 строк и 2 столбцов. Сохраните результат в `table`

В переменную `even` сохраните четные числа от 2 до 16 (включительно)

Скопируйте `even` в переменную `mix`. Перемешайте числа в `mix` так, чтобы массив изменился

Получите из even 3 числа без повторений. Сохраните их в переменную `select`

Получите переменную `triplet`, которая должна содержать перемешанные значения из массива `select` (сам `select` измениться не должен)

In [ ]:
import numpy as np


np.random.seed(2021)

# simple = np.random.rand()
# print(simple)

# randoms = np.random.uniform(-150, 2021, size=(120))
# print(randoms)

# table = np.random.randint(1, 101, size=(3, 2))
# print(table)

even =  np.arange(2,17,2)
print(even)

# mix = even.copy()

# np.random.shuffle(mix)
# print(mix)

# select =  np.random.choice(even, size=3, replace=False)
# print(select)

# triplet = np.random.permutation(select)
# print(triplet)
# print(select)

[ 2  4  6  8 10 12 14 16]


Задание 10.5

Получите сумму чисел, сохранённых в переменных x и y.

In [ ]:
x = 2147483548

y = 2147471336

np.sum(np.int64(x) + np.int64(y))

4294954884

Задание 10.6

Напишите функцию `get_chess(a)`, которая принимает на вход длину стороны квадрата `a` и возвращает двумерный массив формы `(a, a)`, заполненный 0 и 1 в шахматном порядке. В левом верхнем углу всегда должен быть ноль.

Примечание. воспользуйтесь функцией `np.zeros`, а затем с помощью срезов без циклов задайте необходимым элементам значение 1.

В Python для получения каждого второго элемента используется срез [::2]. Подумайте, как грамотно применить этот принцип к двумерному массиву.

Пример


`get_chess(1)`

`# array([[0.]])`

`get_chess(4)`

`# array([[0., 1., 0., 1.],`

`#        [1., 0., 1., 0.],`

`#        [0., 1., 0., 1.],`

`#        [1., 0., 1., 0.]])`

In [ ]:
import numpy as np

def get_chess(a):
    if a > 1:
        zeros_1d = np.zeros((a,a))
        zeros_1d[::2, 1::2] = 1
        zeros_1d[1::2, ::2] = 1
    else:
        zeros_1d = np.zeros((a,a))
    return zeros_1d

get_chess(1)
# array([[0.]])

#get_chess(4)
# array([[0., 1., 0., 1.],
#        [1., 0., 1., 0.],
#        [0., 1., 0., 1.],
#        [1., 0., 1., 0.]])

array([[0.]])

Задание 10.7

Вы разрабатываете приложение для прослушивания музыки. Конечно же, там будет доступна функция перемешивания плейлиста. Пользователю может настолько понравиться перемешанная версия плейлиста, что он захочет сохранить его копию. Однако вы не хотите хранить в памяти новую версию плейлиста, а просто хотите сохранять тот seed, с которым он был сгенерирован.

Для этого напишите функцию `shuffle_seed(array)`, которая принимает на вход массив из чисел, генерирует случайное число для seed в диапазоне от 0 до 2**32 - 1 (включительно) и возвращает кортеж: перемешанный с данным seed массив (исходный массив должен оставаться без изменений), а также seed, с которым этот массив был получен.

Пример


`array = [1, 2, 3, 4, 5]`

`shuffle_seed(array)`

`# (array([1, 3, 2, 4, 5]), 2332342819)`

`shuffle_seed(array)`

`# (array([4, 5, 2, 3, 1]), 4155165971)`

Примечание. Не забудьте импортировать бибилиотеки.

In [ ]:
import numpy as np

array = [1, 2, 3, 4, 5]

def shuffle_seed(array):
    num_random = np.random.randint(2**32-1, dtype=np.uint32)
    np.random.seed(num_random)
    pley_list = np.random.permutation(array)
    return pley_list, num_random

shuffle_seed(array)
# (array([1, 3, 2, 4, 5]), 2332342819)
shuffle_seed(array)
# (array([4, 5, 2, 3, 1]), 4155165971)


(array([1, 3, 2, 4, 5]), 1269965141)

In [ ]:
num_random = np.random.randint((2**32-1), dtype=np.uint32)
num_random

3408972076

In [ ]:
print(2**32-1)
np.iinfo(np.int32)


4294967295


iinfo(min=-2147483648, max=2147483647, dtype=int32)

Задание 10.8

Напишите функцию `min_max_dist(*vectors)`, которая принимает на вход неограниченное число векторов через запятую. Гарантируется, что все векторы, которые передаются, одинаковой длины.

Функция возвращает минимальное и максимальное расстояние между векторами в виде кортежа.

Пример


`vec1 = np.array([1,2,3])`

`vec2 = np.array([4,5,6])`

`vec3 = np.array([7, 8, 9])`

`min_max_dist(vec1, vec2, vec3)`

`# (5.196152422706632, 10.392304845413264)`

In [ ]:
# Вариант решения №1

import numpy as np

vec1 = np.array([1,2,3])

vec2 = np.array([4,5,6])

vec3 = np.array([7, 8, 9])

def min_max_dist(*vectors):
    list_min_max = []
    for i in range(len(vectors)):
        for j in range(i+1, len(vectors)):
            list_min_max.append(np.linalg.norm(vectors[i] - vectors[j]))

    return np.min(list_min_max), np.max(list_min_max)


min_max_dist(vec1, vec2, vec3)

# (5.196152422706632, 10.392304845413264)

(5.196152422706632, 10.392304845413264)

In [ ]:
# Вариант решения №2

import numpy as np

vec1 = np.array([1,2,3])

vec2 = np.array([4,5,6])

vec3 = np.array([7, 8, 9])

def min_max_dist(*vectors):
    list_min_max = []
    for i in vectors:
        for j in vectors:
            if (j == i).any():
                continue
            list_min_max.append(np.linalg.norm(i - j))

    return np.min(list_min_max), np.max(list_min_max)


min_max_dist(vec1, vec2, vec3)

# (5.196152422706632, 10.392304845413264

(5.196152422706632, 10.392304845413264)

Задание 10.9

Напишите функцию `any_normal`, которая принимает на вход неограниченное число векторов через запятую. Гарантируется, что все векторы, которые передаются, одинаковой длины.

Функция возвращает `True`, если есть хотя бы одна пара перпендикулярных векторов. Иначе возвращает `False`.

Пример

vec1 = np.array([2, 1])

vec2 = np.array([-1, 2])

vec3 = np.array([3,4])

print(any_normal(vec1, vec2, vec3))

`# True`

In [ ]:
import numpy as np

vec1 = np.array([2, 1])

vec2 = np.array([-1, 2])

vec3 = np.array([3,4])

def any_normal(*vectors):

    for i in vectors:
        for j in vectors:
            if (j == i).any():
                continue
            if not np.dot(vec1, vec2):
                return True

    return False



print(any_normal(vec1, vec2, vec3))

# True

True


Задание 10.10

Напишите функцию `get_loto(num)`, генерирующую трёхмерный массив случайных целых чисел от 1 до 100 (включительно). Это поля для игры в лото.

Трёхмерный массив должен состоять из таблиц чисел формы 5х5, то есть итоговая форма — (num, 5, 5).

Функция возвращает полученный массив.

Пример

get_loto(3)

Результат:

array(

     [[[ 35,  66,  38,  11,  32],
       [ 32,   7,  37,  83,  42],
       [ 89,  37,  19,  51,  89],
       [ 70, 100,  83,   5,  11],
       [ 20,  13,  60,  26,  41]],

      [[ 23,  49,  76,  44,  43],
       [ 59,  63,  99,  92,   2],
       [ 83,   4,  25,  73,  19],
       [ 10,  18,  40,  11,  21],
       [ 58,  45,  73,  93,  61]],

      [[100,  88,  70,  34,  51],
       [  5,  35,  36,  85,  88],
       [ 72,  23,  87,  30,  40],
       [ 29,  21,  51,  73,  81],
       [ 91,  19,  87,  60,  27]]])

In [ ]:
import numpy as np

def get_loto(num):

    arry_bilet = np.random.randint(1, 101, size=(num,5,5))
    return arry_bilet

get_loto(3)
#array([[[ 35,  66,  38,  11,  32],
       #[ 32,   7,  37,  83,  42],
       #[ 89,  37,  19,  51,  89],
       #[ 70, 100,  83,   5,  11],
       #[ 20,  13,  60,  26,  41]],

      #[[ 23,  49,  76,  44,  43],
       #[ 59,  63,  99,  92,   2],
       #[ 83,   4,  25,  73,  19],
       #[ 10,  18,  40,  11,  21],
       #[ 58,  45,  73,  93,  61]],

      #[[100,  88,  70,  34,  51],
       #[  5,  35,  36,  85,  88],
       #[ 72,  23,  87,  30,  40],
       #[ 29,  21,  51,  73,  81],
       #[ 91,  19,  87,  60,  27]]])

array([[[ 3,  8, 47, 75, 68],
        [75, 14, 55, 69, 99],
        [92, 86, 85, 34, 78],
        [65, 29, 44, 81, 42],
        [44, 25, 59, 73, 67]],

       [[46, 98, 84, 46, 75],
        [42, 55, 25, 16,  6],
        [83, 55, 90, 83, 50],
        [81, 18, 38, 28, 29],
        [68, 70, 58, 84, 50]],

       [[69, 24, 21, 18, 19],
        [58, 63, 89, 46, 37],
        [39, 85, 80, 83, 86],
        [74,  4, 86, 87, 81],
        [62, 45, 64, 48, 52]]])

Задание 10.11

Напишите функцию `get_unique_loto(num)`. Она так же, как и функция в задании 49.9.10, генерирует `num` полей для игры в лото, однако теперь на каждом поле 5х5 числа не могут повторяться.

Функция также должна возвращать массив формы `num` x 5 x 5.

Пример

`get_unique_loto(3)`

Результат:

array(
     
     [[[26, 91, 73,  7, 16],
       [46, 85, 78, 77, 51],
       [84, 86, 55, 71, 58],
       [17, 61, 50, 30, 97],
       [66, 29, 38, 41, 32]],

      [[49, 32,  3, 21, 85],
       [45,  8, 94, 46, 96],
       [41, 38, 58, 37, 98],
       [66, 77, 54, 80, 26],
       [62, 63, 72,  5, 43]],

      [[55, 60,  6, 80, 97],
       [23, 69, 94,  9, 24],
       [17, 98, 27, 63, 79],
       [84, 74, 51, 39, 54],
       [77, 30, 48, 75, 85]]])

In [8]:
import numpy as np

def get_unique_loto(num):

    numbers = np.arange(1, 101)

    array_f = []
    for i in range(num):
        array_f.append(np.random.choice(numbers, size=(5,5), replace=False))

    return np.array(array_f)

get_unique_loto(3)

array([[[80, 29, 13, 95, 74],
        [54, 37,  2, 57, 67],
        [55, 86, 51, 11, 90],
        [30, 47, 56, 63, 84],
        [ 9, 12, 62, 50, 89]],

       [[ 7, 88, 50, 99, 40],
        [32, 77, 12, 93, 68],
        [44, 60, 26, 19, 79],
        [15, 45, 13,  4, 27],
        [96, 41, 20, 30, 37]],

       [[22,  2, 15,  1, 12],
        [75, 87, 17,  5, 67],
        [53, 51, 21, 54, 85],
        [70, 94, 52, 25, 49],
        [56,  4, 82, 44, 16]]])

Объединение массивов

In [16]:
import numpy as np

vec1 = np.array([1,2,3])
vec2 = np.array([4,5,6])
vec3 = np.array([7,8,9])

vec = np.concatenate((vec1, vec2, vec3))
print(vec)

[1 2 3 4 5 6 7 8 9]


Посмотреть размер массива

In [15]:
vec.size

9

In [17]:
simplelist = [19, 242, 14, 152, 142, 1000]
np.mean(simplelist)

261.5